### DISCUSSION TITLE
TinyUSB on x86_64 (AMD and Intel)?

### DISCUSSION THREAD

**Discussion by lonnietc** [hathach/tinyusb#987] [Q&A]
Hello All,

I am working on a new type of hypervisor project (Type-1) and am porting over the MUSL Clib under gcc/g++ to it since it currently has no standard libraries.

In as much, the hypervisor will run on x86_64 (Intel and AMD) processors and I need to add in some USB capabilities where as now the interface for the user is based upon the old PS/2 type for keyboard and mouse, but I want to try and update things so that it could have USB drivers for these and other capabilities.

I have also heard about project "Tinyusb" and "Libusb" which also seems good but I do not know which might have the least OS dependencies since I have minimal standard library support so far.

It seems that Tinyusb can run on bare metal, if I read the documentation correctly, but that seems to be looking for specific boards and I wanted to try and find out if it can be made to run on these x86_64 platforms that I am using.

Still very new to a lot of this development, but I would greatly appreciate any help or guidance that you might be willing to offer.

Thanks in advance and have a good weekend.

**Comment by hathach** [comment]
yeah, TinyUSB is mostly running on bare metal, but so far only on small MCUs. For x86, to be honest, I haven't thought x86 would ever need the help from TinyUSB. Though, it is still possible, I guess, people did get TinyUSB running on bare metal with Zynq 7000.
- TinyUSB already support EHCI (USB 2.0), though EHCI controller includes some of NXP-specific tweak for period scheduler, but can be easy to get it running with standard one. Note: there is no support for xHCI though. 
- HID keyboard/mouse are already supported so yeah, it should work well enough.
- The stock example look for the specific board since each mcu and each boards has its own set of peripherals, gpio, clock setup that need to be init-ed correctly to work. But it isn't to complicated to add, you may find https://github.com/hathach/tinyusb/blob/master/docs/boards.md#add-your-own-board and https://github.com/hathach/tinyusb/blob/master/docs/getting_started.md useful to look at.

**Comment by lonnietc** [comment]
@hathach Thanks for your response and I will look into it although I have only minimal experience in this area.

Is there some tutorial around somewhere that you might be aware of that can discuss things like what exactly the xHCI is along with other components since I need to get a better understanding of the whole process and how it all comes together.

I will look more at the links that you have provided as well.

Cheers

**Reply by hathach** [reply]
so far only OHCI & EHCI are supported by TinyUSB, since there is no xHCI controller in any MCU that I know. If your bsp only have xHCI, you will need to write your own driver to hook up with the stack. There is an hcd API that you have to implement https://github.com/hathach/tinyusb/blob/master/src/host/hcd.h 

The documentation of the project isn't too great, If you aren't familiar with MCU, you may end up reading the code, which can be a bit overwhelmed at first. I would suggest you to get a couple of 4$ raspberrypi pico board to start with, Pi co has a great documentation to get started, and you could run the hid keyboard/mouse example with their MCU. The application API to run on x86 is ideally the same as one running on other MCU (in this case is pico). You could try out tinyusb to see if you like it or not before invest your time and effort to the port. 

https://www.youtube.com/watch?v=2eB8p8vo7lw
